File: lib/java/src/main/java/io/github/brenbar/japi/SyncClient.java
package io.github.brenbar.japi;

import java.util.*;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;

import io.github.brenbar.japi.Client.Middleware;
import io.github.brenbar.japi.Client.ModifyHeaders;

public class SyncClient {

    private Client client;
    private Function<byte[], Future<byte[]>> syncTransport;
    private Serializer serializer;
    private Long timeoutMs;

    public SyncClient(Function<byte[], Future<byte[]>> syncTransport) {
        this.client = new Client(this::serializeAndTransport);
        this.syncTransport = syncTransport;
        this.serializer = new DefaultSerializer();
        this.timeoutMs = 5000L;
    }

    public SyncClient setSerializer(Serializer serializer) {
        this.serializer = serializer;
        return this;
    }

    public SyncClient setTimeoutMs(Long timeoutMs) {
        this.timeoutMs = timeoutMs;
        return this;
    }

    public SyncClient setModifyHeaders(ModifyHeaders modifyHeaders) {
        client.modifyHeaders = modifyHeaders;
        return this;
    }

    public SyncClient setMiddleware(Middleware middleware) {
        client.middleware = middleware;
        return this;
    }

    public SyncClient setUseBinary(boolean useBinary) {
        client.useBinary = useBinary;
        return this;
    }

    public SyncClient setForceSendJson(boolean forceSendJson) {
        client.forceSendJson = forceSendJson;
        return this;
    }

    public Map<String, Object> submit(
            Request request) {
        return client.submit(request);
    }

    protected List<Object> serializeAndTransport(List<Object> inputMessage, boolean sendAsMsgPack) {
        try {
            var headers = (Map<String, Object>) inputMessage.get(1);
            headers.put("_tim", timeoutMs);

            byte[] inputMessageBytes = InternalClientProcess.serialize(inputMessage, serializer,
                    sendAsMsgPack);

            var outputMessageBytes = syncTransport.apply(inputMessageBytes).get(this.timeoutMs,
                    TimeUnit.MILLISECONDS);

            List<Object> outputMessage = InternalClientProcess.deserialize(outputMessageBytes, serializer);
            return outputMessage;
        } catch (Exception e) {
            throw new ClientProcessError(e);
        }
    }
}

File: lib/py/japi/sync_client.py
from typing import List, Dict, Any, Callable, Union
import concurrent.futures

from japi.client_process_error import ClientProcessError
from japi.client import Client
from japi.default_serializer import DefaultSerializer
from japi.serializer import Serializer


class SyncClient(Client):

    def __init__(self, sync_transport: Callable[[bytes], concurrent.futures.Future[bytes]], timeout_ms: int = 5000, serializer: Serializer = DefaultSerializer(), use_binary: bool = False, force_send_json: bool = True):
        super().__init__(use_binary=use_binary, force_send_json=force_send_json)
        self.sync_transport = sync_transport
        self.serializer = serializer
        self.timeout_ms = timeout_ms

    def serialize_and_transport(
        self, input_japi_message: List[Any], use_msg_pack: bool
    ) -> List[Any]:
        try:
            if use_msg_pack:
                input_japi_message_payload = self.serializer.serialize_to_msg_pack(
                    input_japi_message
                )
            else:
                input_japi_message_payload = self.serializer.serialize_to_json(
                    input_japi_message
                )

            output_japi_message_payload = self.sync_transport(
                input_japi_message_payload
            ).result(self.timeout_ms / 1000)

            output_japi_message: List[Any]
            if output_japi_message_payload[0] == ord('['):
                output_japi_message = self.serializer.deserialize_from_json(
                    output_japi_message_payload
                )
            else:
                output_japi_message = self.serializer.deserialize_from_msg_pack(
                    output_japi_message_payload
                )

            return output_japi_message
        except Exception as e:
            raise ClientProcessError(e)
